content/solids/components/voxel-painter.ts

```diff
@@ -0,0 +1,462 @@
+        const isInShapeFunction = isInShapeFunctions[shape]
+        const p = new THREE.Vector3()
+        for (let i = -10; i < 10; i++)
+          for (let j = -10; j < 10; j++)
+            for (let k = -10; k < 10; k++)
+            {
+              p.set(i, j, k).addScalar(.5)
+              if (isInShapeFunction(p)) {
+                let voxel = Voxel()
+                voxel.position.copy(p)
+              }
+            }
+      }
+
+      const coords = shape.split(",")
+      for (let i = 0, il = Math.floor(coords.length / 3); i < il; ++i)
+      {
+        const voxel = Voxel()
+        voxel.position.set(
+          parseFloat(coords[i * 3 + 0]),
+          parseFloat(coords[i * 3 + 1]),
+          parseFloat(coords[i * 3 + 2]))
+        snapToNearestValidCubeCenterPosition(voxel.position)
+      }
+    }
+
+    var raycaster = new THREE.Raycaster();
+    var oldRay = new THREE.Ray()
+
+    //can be rotating, placing, erasing, null
+    let mouseControlMode = ""
+
+    const placingStart = new THREE.Vector3()
+    const placingEnd = new THREE.Vector3()
+    {
+      const gridDimension = 20
+      scene.add(new THREE.GridHelper(gridDimension,gridDimension));
+      
+      var floorIntersectionPlaneGeometry = new THREE.PlaneBufferGeometry(gridDimension, gridDimension);
+      floorIntersectionPlaneGeometry.rotateX(-Math.PI / 2);
+      var floorIntersectionPlane = new THREE.Mesh(floorIntersectionPlaneGeometry, new THREE.MeshBasicMaterial({ visible: false }));
+      scene.add(floorIntersectionPlane);
+      objectsOnWhichVoxelsCanBePlaced.push(floorIntersectionPlane);
+
+      var placingHelpers = Array()
+      for(let i = 0; i < 3; i++)
+      {
+        const placingHelper = new THREE.Mesh(new THREE.PlaneBufferGeometry(999., 999.), new THREE.MeshBasicMaterial({ side: THREE.DoubleSide, 
+          // transparent:true,
+          // opacity:.5,
+          visible:false 
+        }))
+        placingHelpers.push(placingHelper)
+        scene.add(placingHelper)
+      }
+      placingHelpers[0].geometry.rotateX(-TAU / 4.) //shouldn't need negative, check both directions angles
+      placingHelpers[1].geometry.rotateY( TAU / 4.)
+      placingHelpers[2].geometry.rotateZ( TAU / 4.)
+    }
+
+    function snapToNearestValidCubeCenterPosition(p) { p.floor().addScalar(.5) }
+
+    function setPositionFromVoxelIntersection(p,intersection)
+    {
+      p.copy(intersection!.face!.normal).multiplyScalar(.1)
+      p.add(intersection.point)
+      snapToNearestValidCubeCenterPosition(p)
+    }
+
+    let asyncRaycaster = new THREE.Raycaster()
+    $canvas.on('pointermove', (event: PointerEvent) => {
+      event.preventDefault();
+      
+      const p = canvasPointerPosition(event, $canvas);
+      camera.updateMatrixWorld()
+      v1.set((p.x / $canvas.width ) * 2. - 1.,
+            -(p.y / $canvas.height) * 2. + 1.);
+      asyncRaycaster.setFromCamera(v1, camera);
+    });
+
+    $canvas.on('pointerdown', (event: PointerEvent) => {
```

Comment:
use
```
slide($canvas, {
  start: () => {},
  move: () => {},
  end: () => {}
})
```
to handle slide events (petter performance + accessibility)

Reply:
Done